crates/admin/src/storage_query/query.rs

```diff
@@ -58,125 +57,147 @@ pub struct QueryRequest {
-        };
+#[derive(Clone)]
+// unfortunately the json writer doesnt give a way to get a mutable reference to the underlying writer, so we need another pointer in to its buffer
+// we use a lock here to help make the writer send/sync, despite it being totally uncontended :(
+struct LockWriter(Arc<RwLock<Vec<u8>>>);
+
+impl LockWriter {
+    fn new() -> Self {
+        Self(Arc::new(RwLock::new(Vec::new())))
+    }
+
+    fn take(&self) -> Vec<u8> {
+        let mut vec = self.0.write().unwrap();
+        let new_vec = Vec::with_capacity(vec.capacity());
+        std::mem::replace(&mut vec, new_vec)
     }
-    RecordBatch::try_new(converted_schema, columns)
 }
 
-fn convert_array_offset<Before: ByteArrayType, After: ByteArrayType>(
-    array: &GenericByteArray<Before>,
-) -> Result<GenericByteArray<After>, ArrowError>
-where
-    After::Offset: TryFrom<Before::Offset>,
-{
-    let offsets = array
-        .offsets()
-        .iter()
-        .map(|&o| After::Offset::try_from(o))
-        .collect::<Result<ScalarBuffer<After::Offset>, _>>()
-        .map_err(|_| ArrowError::CastError("offset conversion failed".into()))?;
-    GenericByteArray::<After>::try_new(
-        OffsetBuffer::new(offsets),
-        array.values().clone(),
-        array.nulls().cloned(),
-    )
+impl Write for LockWriter {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.0.write().unwrap().write(buf)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.0.write().unwrap().flush()
+    }
 }
 
-/// Convert the record batches so that they don't contain LargeUtf8 or LargeBinary columns as JS doesn't
-/// support these yet.
-struct ConvertRecordBatchStream {
+struct JsonWriter {
+    json_writer: datafusion::arrow::json::Writer<LockWriter, JsonArray>,
+    lock_writer: LockWriter,
+    started: bool,
+    finished: bool,
+}
+
+impl RecordBatchWriter for JsonWriter {
+    fn new(_schema: &Schema) -> Result<Self, DataFusionError> {
+        let mut lock_writer = LockWriter::new();
+        // we write out under 'rows' key so that we may add extra keys later (eg 'schema')
+        lock_writer.write_all(br#"{"rows":"#)?;
+        Ok(Self {
+            json_writer: datafusion::arrow::json::Writer::new(lock_writer.clone()),
+            lock_writer,
+            started: false,
+            finished: false,
+        })
+    }
+
+    fn write(&mut self, batch: &RecordBatch) -> Result<Bytes, DataFusionError> {
+        self.started = true;
+        self.json_writer.write(batch)?;
+        Ok(Bytes::from(self.lock_writer.take()))
+    }
+
+    fn finish(&mut self) -> Result<Bytes, DataFusionError> {
+        if !self.finished {
+            self.finished = true;
+            if self.started {
+                self.lock_writer.write_all(b"]}")?;
+            } else {
+                self.lock_writer.write_all(b"[]}")?;
+            }
+        }
+        Ok(Bytes::from(self.lock_writer.take()))
+    }
+}
+
+#[pin_project]
```

Comment:
It took me a while to understand why you are pin projecting the `WriteRecordBatchStream` because you never call `project` or `project_ref` in the `Stream` implementation. I think you use it to make use of a conditional `Unpin` implementation for the `WriteRecordBatchStream` which is generated by the `pin_project` macro. I think this is quite subtle and hard to understand. Instead, I would suggest to add the `+ Unpin` bound to `W` in line 200.

Reply:
Ah, I am not sure I knew myself how it worked. That makes sense thanks